File: Caster/Interface/Themes/PlusThemes.swift
import SwiftUI
import Observation

/// The palette the app is wearing.
///
/// `.system` is the free default and follows the phone's light/dark setting, as
/// the app always has. The named ones are Plus, and each is a fixed look rather
/// than a light/dark pair — picking "Midnight" is choosing a dark app, not
/// choosing how a dark app looks.
///
/// Every one of them keeps `PlayerPalette.colors` for the seats. Seat colour is
/// identity, not decoration: it is how a person finds their own finger on a
/// crowded screen, it is pinned by `shared/parity/golden.json` so the two
/// platforms agree, and a theme that recoloured it would change how the games
/// *play*. Themes dress the chrome and nothing else.
enum ThemeSelection: String, CaseIterable, Identifiable, Codable {
    case system
    case midnight
    case dusk
    case forest
    case paper

    var id: String { rawValue }

    var isPlus: Bool { self != .system }

    var title: String {
        switch self {
        case .system: return "System"
        case .midnight: return "Midnight"
        case .dusk: return "Dusk"
        case .forest: return "Forest"
        case .paper: return "Paper"
        }
    }

    var subtitle: String {
        switch self {
        case .system: return "Follows your phone"
        case .midnight: return "Near-black, high contrast"
        case .dusk: return "Warm plum and amber"
        case .forest: return "Deep green, low glare"
        case .paper: return "Off-white and ink"
        }
    }

    func palette(for scheme: ColorScheme) -> Theme {
        switch self {
        case .system: return Theme.forScheme(scheme)
        case .midnight: return PlusPalettes.midnight
        case .dusk: return PlusPalettes.dusk
        case .forest: return PlusPalettes.forest
        case .paper: return PlusPalettes.paper
        }
    }

    /// The two swatches the settings row shows, so a palette can be judged
    /// without applying it.
    func swatch(for scheme: ColorScheme) -> (background: Color, accent: Color) {
        let theme = palette(for: scheme)
        return (theme.background, theme.accent)
    }
}

/// The four Plus palettes. Each is one `Theme` value of the same eleven fields
/// as `LightTheme` and `DarkTheme` — there is no new machinery here, which is
/// what makes these cheap enough to be worth shipping.
enum PlusPalettes {
    static let midnight = Theme(
        background: Color(red: 0.008, green: 0.012, blue: 0.024),
        surface: Color(red: 0.024, green: 0.031, blue: 0.051),
        surfaceRaised: Color(red: 0.055, green: 0.071, blue: 0.106),
        border: Color(red: 0.129, green: 0.157, blue: 0.212),
        textPrimary: Color(red: 0.925, green: 0.945, blue: 0.980),
        textSecondary: Color(red: 0.529, green: 0.573, blue: 0.647),
        accent: Color(red: 0.318, green: 0.627, blue: 0.996),
        success: Color(red: 0.204, green: 0.827, blue: 0.600),
        danger: Color(red: 0.984, green: 0.404, blue: 0.447),
        warning: Color(red: 0.984, green: 0.780, blue: 0.353),
        playerColors: PlayerPalette.colors
    )

    static let dusk = Theme(
        background: Color(red: 0.086, green: 0.043, blue: 0.098),
        surface: Color(red: 0.129, green: 0.067, blue: 0.145),
        surfaceRaised: Color(red: 0.192, green: 0.102, blue: 0.208),
        border: Color(red: 0.310, green: 0.180, blue: 0.325),
        textPrimary: Color(red: 0.984, green: 0.949, blue: 0.965),
        textSecondary: Color(red: 0.706, green: 0.588, blue: 0.694),
        accent: Color(red: 0.976, green: 0.596, blue: 0.310),
        success: Color(red: 0.353, green: 0.812, blue: 0.596),
        danger: Color(red: 0.965, green: 0.400, blue: 0.482),
        warning: Color(red: 0.988, green: 0.808, blue: 0.400),
        playerColors: PlayerPalette.colors
    )

    static let forest = Theme(
        background: Color(red: 0.031, green: 0.086, blue: 0.067),
        surface: Color(red: 0.047, green: 0.125, blue: 0.098),
        surfaceRaised: Color(red: 0.075, green: 0.180, blue: 0.141),
        border: Color(red: 0.145, green: 0.290, blue: 0.235),
        textPrimary: Color(red: 0.925, green: 0.965, blue: 0.941),
        textSecondary: Color(red: 0.573, green: 0.694, blue: 0.635),
        accent: Color(red: 0.443, green: 0.827, blue: 0.502),
        success: Color(red: 0.325, green: 0.847, blue: 0.588),
        danger: Color(red: 0.937, green: 0.435, blue: 0.396),
        warning: Color(red: 0.933, green: 0.780, blue: 0.365),
        playerColors: PlayerPalette.colors
    )

    static let paper = Theme(
        background: Color(red: 0.976, green: 0.965, blue: 0.933),
        surface: Color(red: 0.988, green: 0.980, blue: 0.957),
        surfaceRaised: Color(red: 0.937, green: 0.918, blue: 0.871),
        border: Color(red: 0.859, green: 0.831, blue: 0.769),
        textPrimary: Color(red: 0.114, green: 0.106, blue: 0.086),
        textSecondary: Color(red: 0.396, green: 0.373, blue: 0.325),
        accent: Color(red: 0.706, green: 0.325, blue: 0.184),
        success: Color(red: 0.216, green: 0.518, blue: 0.353),
        danger: Color(red: 0.729, green: 0.204, blue: 0.204),
        warning: Color(red: 0.729, green: 0.510, blue: 0.114),
        playerColors: PlayerPalette.colors
    )
}

/// Remembers the chosen palette. Its own tiny store rather than a field on
/// `AppEnvironment`, for the same reason the wheels have one: a preference that
/// lives in memory is a preference that a relaunch throws away.
///
/// Falls back to `.system` whenever Plus is not held, so a refund or a restore
/// on a new device can never strand somebody in a palette they no longer own —
/// without forgetting which one they had picked, in case they buy it again.
@Observable
@MainActor
final class ThemeStore {
    private static let storageKey = "caster.theme.selection"

    private(set) var preferred: ThemeSelection = .system

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        if let raw = defaults.string(forKey: Self.storageKey),
           let stored = ThemeSelection(rawValue: raw) {
            preferred = stored
        }
    }

    func select(_ selection: ThemeSelection) {
        preferred = selection
        defaults.set(selection.rawValue, forKey: Self.storageKey)
    }

    func effective(hasPlus: Bool) -> ThemeSelection {
        (preferred.isPlus && !hasPlus) ? .system : preferred
    }
}

File: android/app/src/main/java/com/jesperhaafkes/caster/ui/theme/PlusThemes.kt
package com.jesperhaafkes.caster.ui.theme

import android.content.SharedPreferences
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.graphics.Color

/**
 * The palette the app is wearing.
 *
 * [SYSTEM] is the free default and follows the phone's light/dark setting, as
 * the app always has. The named ones are Plus, and each is a fixed look rather
 * than a light/dark pair — picking "Midnight" is choosing a dark app, not
 * choosing how a dark app looks.
 *
 * Every one of them keeps [PlayerPalette.colors] for the seats. Seat colour is
 * identity, not decoration: it is how a person finds their own finger on a
 * crowded screen, it is pinned by `shared/parity/golden.json` so the two
 * platforms agree, and a theme that recoloured it would change how the games
 * *play*. Themes dress the chrome and nothing else.
 *
 * The twin of `ThemeSelection` in `Caster/Interface/Themes/PlusThemes.swift`.
 * The [key] values are the Swift enum's raw values, so the two apps write the
 * same string for the same palette.
 */
enum class ThemeSelection(
    val key: String,
    val title: String,
    val subtitle: String,
) {
    SYSTEM("system", "System", "Follows your phone"),
    MIDNIGHT("midnight", "Midnight", "Near-black, high contrast"),
    DUSK("dusk", "Dusk", "Warm plum and amber"),
    FOREST("forest", "Forest", "Deep green, low glare"),
    PAPER("paper", "Paper", "Off-white and ink");

    val isPlus: Boolean get() = this != SYSTEM

    /** [systemPalette] is what SYSTEM resolves to — light or dark, live. */
    fun palette(systemPalette: Theme): Theme = when (this) {
        SYSTEM -> systemPalette
        MIDNIGHT -> PlusPalettes.midnight
        DUSK -> PlusPalettes.dusk
        FOREST -> PlusPalettes.forest
        PAPER -> PlusPalettes.paper
    }

    companion object {
        fun fromKey(key: String?): ThemeSelection =
            entries.firstOrNull { it.key == key } ?: SYSTEM
    }
}

/**
 * The four Plus palettes. Each is one [Theme] value of the same eleven fields
 * as [LightTheme] and [DarkTheme] — there is no new machinery here, which is
 * what makes these cheap enough to be worth shipping.
 *
 * The numbers are the same ones in `PlusThemes.swift`. They are not checked by
 * `golden.json` because chrome colour is not game feel, but they are meant to
 * match, and a palette changed on one platform should be changed on both.
 */
object PlusPalettes {
    val midnight = Theme(
        background = Color(0.008f, 0.012f, 0.024f),
        surface = Color(0.024f, 0.031f, 0.051f),
        surfaceRaised = Color(0.055f, 0.071f, 0.106f),
        border = Color(0.129f, 0.157f, 0.212f),
        textPrimary = Color(0.925f, 0.945f, 0.980f),
        textSecondary = Color(0.529f, 0.573f, 0.647f),
        accent = Color(0.318f, 0.627f, 0.996f),
        success = Color(0.204f, 0.827f, 0.600f),
        danger = Color(0.984f, 0.404f, 0.447f),
        warning = Color(0.984f, 0.780f, 0.353f),
        playerColors = PlayerPalette.colors,
    )

    val dusk = Theme(
        background = Color(0.086f, 0.043f, 0.098f),
        surface = Color(0.129f, 0.067f, 0.145f),
        surfaceRaised = Color(0.192f, 0.102f, 0.208f),
        border = Color(0.310f, 0.180f, 0.325f),
        textPrimary = Color(0.984f, 0.949f, 0.965f),
        textSecondary = Color(0.706f, 0.588f, 0.694f),
        accent = Color(0.976f, 0.596f, 0.310f),
        success = Color(0.353f, 0.812f, 0.596f),
        danger = Color(0.965f, 0.400f, 0.482f),
        warning = Color(0.988f, 0.808f, 0.400f),
        playerColors = PlayerPalette.colors,
    )

    val forest = Theme(
        background = Color(0.031f, 0.086f, 0.067f),
        surface = Color(0.047f, 0.125f, 0.098f),
        surfaceRaised = Color(0.075f, 0.180f, 0.141f),
        border = Color(0.145f, 0.290f, 0.235f),
        textPrimary = Color(0.925f, 0.965f, 0.941f),
        textSecondary = Color(0.573f, 0.694f, 0.635f),
        accent = Color(0.443f, 0.827f, 0.502f),
        success = Color(0.325f, 0.847f, 0.588f),
        danger = Color(0.937f, 0.435f, 0.396f),
        warning = Color(0.933f, 0.780f, 0.365f),
        playerColors = PlayerPalette.colors,
    )

    val paper = Theme(
        background = Color(0.976f, 0.965f, 0.933f),
        surface = Color(0.988f, 0.980f, 0.957f),
        surfaceRaised = Color(0.937f, 0.918f, 0.871f),
        border = Color(0.859f, 0.831f, 0.769f),
        textPrimary = Color(0.114f, 0.106f, 0.086f),
        textSecondary = Color(0.396f, 0.373f, 0.325f),
        accent = Color(0.706f, 0.325f, 0.184f),
        success = Color(0.216f, 0.518f, 0.353f),
        danger = Color(0.729f, 0.204f, 0.204f),
        warning = Color(0.729f, 0.510f, 0.114f),
        playerColors = PlayerPalette.colors,
    )
}

/**
 * Remembers the chosen palette. Its own tiny store rather than a field on
 * `AppEnvironment`, for the same reason the wheels have one: a preference that
 * lives in memory is a preference that a relaunch throws away.
 *
 * Falls back to [ThemeSelection.SYSTEM] whenever Plus is not held, so a refund
 * or a reinstall on a new device can never strand somebody in a palette they no
 * longer own — without forgetting which one they had picked, in case they buy
 * it again.
 */
class ThemeStore(private val prefs: SharedPreferences) {

    var preferred: ThemeSelection by mutableStateOf(ThemeSelection.SYSTEM)
        private set

    init {
        preferred = ThemeSelection.fromKey(prefs.getString(STORAGE_KEY, null))
    }

    fun select(selection: ThemeSelection) {
        preferred = selection
        prefs.edit().putString(STORAGE_KEY, selection.key).apply()
    }

    fun effective(hasPlus: Boolean): ThemeSelection =
        if (preferred.isPlus && !hasPlus) ThemeSelection.SYSTEM else preferred

    private companion object {
        const val STORAGE_KEY = "caster.theme.selection"
    }
}
